### ISSUE TITLE
O(n^2) performance regression between v0.3.23 and v0.4.0 on long file

### ISSUE THREAD

**Issue by dkegel-fastly** [dominikh/go-tools#1407]
- The output of 'staticcheck -version'
staticcheck 2023.1.3 (v0.4.3)

- The output of 'staticcheck -debug.version' (it is fine if this command fails)
```
staticcheck 2023.1.3 (v0.4.3)

Compiled with Go version: go1.19.6
Main module:
        honnef.co/go/tools@v0.4.3 (sum: h1:o/n5/K5gXqk8Gozvs2cnL0F2S1/g1vcGCAx2vETjITw=)
Dependencies:
        github.com/BurntSushi/toml@v1.2.1 (sum: h1:9F2/+DoOYIOksmaJFPw1tGFy1eDnIJXg+UHjuD8lTak=)
        golang.org/x/exp/typeparams@v0.0.0-20221208152030-732eee02a75a (sum: h1:Jw5wfR+h9mnIYH+OtGT2im5wV1YGGDora5vTv/aa5bE=)
        golang.org/x/mod@v0.7.0 (sum: h1:LapD9S96VoQRhi/GrNTqeBJFrUjs5UHCAtTlgwA5oZA=)
        golang.org/x/sys@v0.3.0 (sum: h1:w8ZOecv6NaNa/zC8944JTU3vz4u6Lagfk4RPQxv92NQ=)
        golang.org/x/tools@v0.4.1-0.20221208213631-3f74d914ae6d (sum: h1:9ZNWAi4CYhNv60mXGgAncgq7SGc5qa7C8VZV8Tg7Ggs=)
```

- The output of 'go version'
go version go1.19.6 darwin/amd64

- The output of 'go env'
GO111MODULE=""
GOARCH="amd64"
GOBIN=""
GOCACHE="/Users/dkegel/Library/Caches/go-build"
GOENV="/Users/dkegel/Library/Application Support/go/env"
GOEXE=""
GOEXPERIMENT=""
GOFLAGS=""
GOHOSTARCH="amd64"
GOHOSTOS="darwin"
GOINSECURE=""
GOMODCACHE="/Users/dkegel/go/pkg/mod"
GONOPROXY="github.com/fastly,github.com/signalsciences"
GONOSUMDB="github.com/fastly,github.com/signalsciences"
GOOS="darwin"
GOPATH="/Users/dkegel/go"
GOPROXY="https://proxy.golang.org,direct"
GOROOT="/usr/local/go"
GOSUMDB="sum.golang.org"
GOTMPDIR=""
GOTOOLDIR="/usr/local/go/pkg/tool/darwin_amd64"
GOVCS=""
GOVERSION="go1.19.6"
GCCGO="gccgo"
GOAMD64="v1"
AR="ar"
CC="clang"
CXX="clang++"
CGO_ENABLED="1"
GOMOD="/Users/dkegel/go.mod"
GOWORK=""
CGO_CFLAGS="-g -O2"
CGO_CPPFLAGS=""
CGO_CXXFLAGS="-g -O2"
CGO_FFLAGS="-g -O2"
CGO_LDFLAGS="-g -O2"
PKG_CONFIG="pkg-config"
GOGCCFLAGS="-fPIC -arch x86_64 -m64 -pthread -fno-caret-diagnostics -Qunused-arguments -fmessage-length=0 -fdebug-prefix-map=/var/folders/v0/0k9hwftd29b18z9z7s2g2sfc0000gn/T/go-build766679827=/tmp/go-build -gno-record-gcc-switches -fno-common"

- Exactly which command you ran
staticcheck

- Output of the command and what's wrong with the output
Takes 12 seconds with 30,000 lines of input; takes 68 seconds with 60,000 lines of input.
Actual file has 700,000 lines of input; version 0.3.22 was usable, current version is not on this file.

- Where we can read the code you're running Staticcheck on
```
package foo

type FooRec struct {
        Yabba string
}

var FooData = []FooRec{
        {"10.20.100.0"},
     ...
}
```

**Comment by dankegel** [comment]
Yes, it is a bit odd to have such a large array in source.  Shouldn't hit this in many apps!

**Comment by dkegel-fastly** [comment]
Yes, it is a bit odd to compile such a big array.  This was for a benchmark.
